:book:  Add blog post for Fake Node Scalability on Metal3 (#500)
metal3-io/metal3-io.github.io: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

- Ironic Python Agent (IPA) is the software that runs on each bare-metal node to carry out Ironic’s instructions. It interacts with the hardware directly, like wiping disks, configuring networks, and handling boot processes.

__Quick Intro: How Ironic and IPA Work Together in Metal3__
Ironic is the bare-metal provisioning tool used in Metal3. Think of it as the system that manages actual servers—turning them on, installing operating systems, and so on. When you want to deploy a Kubernetes cluster on bare-metal hardware, Ironic is the one making sure that each physical server is up and running with the right setup.

The Ironic Python Agent (IPA) plays a super important role here. It’s a little program that runs on the bare-metal machines Ironic is provisioning. You can think of IPA as the "hands" of Ironic—it actually touches the hardware to do things like:

- Booting up the machine.
- Installing the operating system.
- Configuring network settings.

> In a normal Metal3 workflow, BMO translates the kubernetes reconciling logic to concrete actions and forward them to Ironic, which tells the IPA what to do. For example, BMO might notice that  Ironic might tell IPA to boot up a server and install an image, and IPA will carry out these tasks. Once the node is ready, Ironic passes control back to CAPI and CAPM3, which can then set up Kubernetes clusters on these provisioned machines.

Reviewer: ```suggestion
In a normal Metal3 workflow, BMO translates the kubernetes reconciling logic to concrete actions and forward them to Ironic, which as part of the provisioning process tells the IPA what to do. Once the BMH is provisioned, Ironic will only monitor the power state of the machine and IPA will stop running right before the BMH switches to "provisioned" state, CAPI can then set up Kubernetes clusters on these provisioned machines.
```

## The Experiment: Simulating 1000 Kubernetes Clusters
The goal of this experiment was to push Metal3 to simulate 1000 single-node Kubernetes clusters on fake hardware. Instead of provisioning real machines, we used Fake Ironic Python Agents (Fake IP) and Fake Kubernetes API Servers (FKAS) to simulate nodes and control planes, respectively. This setup allowed us to test a massive environment without the need for physical infrastructure.

### Step 1: Setting Up the environment

> As you may have known, a typical Metal3 workflow requires existence of several components: bootstrap kubernetes cluster, special networks, baremetal nodes, etc. As we are working on simulating environment, we will start with a newly spawn Ubuntu VM, create a cluster with minikube, add networks with libvirt, and so on (If you're familiar with the Metal3's dev-env, this step is similar to what script [01](), [02] and a part of [03] do). We will not go into details on this part, but you can find the related setup from [this script]() if you are interested.

Reviewer: "special networks", not really, we use separate provisioning network but it is not a hard requirement.

Author: I think ironic requires `baremetal` network, doesn't it?

Reviewer: e.g in a scenario where we use refish with virtual media booting and there is no network seapration between the NICs including the BMC NIC we can use 1 network.

### Step 2: Install BMO and Ironics

In Metal3’s typical workflow, we usually rely on **Kustomize** to install Ironic and BMO. Kustomize helps us define configurations for Kubernetes resources, making it easier to customize and deploy services. However, the current Kustomize overlay we use for Metal3 is designed to configure **only a single Ironic instance**. This setup works well for smaller environments, but it becomes a bottleneck when you’re trying to scale up and handle thousands of nodes.

That’s where Ironic’s **special mode** comes into play. Ironic has the ability to run **multiple Ironic conductors** while sharing the same database. The best part? Workload balancing between conductors happens automatically. This means that no matter which Ironic conductor receives a request, the load is evenly distributed across all conductors, ensuring efficient provisioning. To achieve this requires separation of **ironic conductor** from the database, from which the conductor part can be scaled up. Each **conductor** will have its own `PROVISIONING_IP`, hence the need of having a specialized `configMap`.

> We used [Helm]() for this purpose. In our Helm chart, **Ironic conductor** container and **HTTP server (httpd)** container are separated into a new pod, while the rest of the ironic package (mostly `mariadb` - ironic database) stays in another pod. A list of `PROVISIONING_IP`s is provided by the chart's `values.yaml`, and for each ip, an  **ironic conductor** pod is created, along with a configmap whose values rendered with the ip's value. This way, we can dynamically scale up/down ironic (or more specifically **ironic conductors** by simply adding/removing ips.

Reviewer: Missing link for Helm, or the link syntax is not needed.

Author: Yeah. As this a WIP, I was writing down markdown link syntax just as placeholders. No proper links were added. Probably I could add the TODO to remember adding them later.

BMO, on the other hand, still works well with kustomize, as we do not need to scale it. As with normal metal3 workflow, BMO and Ironic will need to share some authentication, so that they can work together with TLS.

> You can check out the full Ironic helm chart [here]()

Reviewer: Other missing  link

### Step 2: Creating Fake Nodes with Fake Ironic Python Agents
As we mentioned at the beginning, instead of using real hardware, we will use a new tool called **Fake Ironic Python Agent**, or **Fake IPA** to simulate the nodes.

> Setting up **Fake IPA** is quite straight-forward, as **Fake IPA** runs as containers in the host machine, but first, we need to create the list of "nodes" that we will use (Fake IPA requires to have that list ready when it starts). A "node" typically looks like this

Reviewer: on the host machine instead of in the host machine

```shell
'{
      "uuid": $uuid,
      "name": $node_name,
      "power_state": "Off",
      "external_notifier": "True",
      "nics": [
	{"mac": $macaddr, "ip": "192.168.0.100"}
```

> All of the variables (`uuid`, `node_name`, `macaddress`) can be dynamically generated in any way you want (check [this script]() out if you need an idea), but we will need to store these information in so that we can generate the BMH objects that match those "nodes". The `ip` is, on other hand, not important. It could be anything.

Reviewer: missing link for the script

about ALL the nodes, we can decide what container to locate the "node". In general, it's a good idea to spread them out, so all containers are loaded equally.

We can go ahead and run `kubectl apply -f` on one (or all of) the BMH manifests. What you expect to see is that a BMH object is created, and its state will change from `registering` to `available` after awhile. It means **ironic** acknowledged that the node is valid, in good state and ready to be provisioned.

### Step 4: Deploy the fake nodes to kubernetes clusters

In a normal Metal3 workflow, after having the BMH objects in `available` state, we can go ahead and provision new kubernetes clusters with `clusterctl`. However, with fake nodes, things get a tiny bit more complex. At the end of the provisioning process, **Cluster API** expects that there is a new kubernetes API server created for the new cluster, from which it will check if all nodes are up, all the control planes have `apiserver`, `etcd`, etc. pods up and running, and so on. It is where the [**Fake Kubernetes API Server** (FKAS)]() comes in.

> As the **FKAS README** linked above already described how it works, we won't go into details. Basically, we will just need to send **FKAS** a `register` request, and it will give us an IP and a port, which we can plug into our cluster template, and then run `clusterctl generate cluster`.

Reviewer: I would suggest getting rid of languge like "we won't , we will just ..."   It is just my personal opinion but usually in technical writings this personal tone is improper. But I know in blog posts even in technical topics folks use it. 

Up to you, I would not use it.

Author: Thanks Adam. That's a good advice, the reason I used a rather casual language was to make this post a bit friendlier to newcomers (Same reason why I went on and on about explaining how Metal3, ironic, etc. work), but yeah, I agree that those words aren't always needed.

---
title: "Scaling Kubernetes with Metal3: Simulating 1000 Clusters with Fake Ironic Agents"
date: 2024-10-24
draft: false
categories: ["metal3", "cluster API", "ironic", "baremetal", "scaling"]
author: Huy Mai
---

> ## Introduction

Reviewer 2: I suggest dropping this heading. It is technically directly below the title, so it is two headings on top of each other. That is not a big deal but it also makes the excerpt look strange since there it isn't visible as a heading. It looks like this:

> Introduction If you’ve ever tried scaling out Kubernetes clusters in a bare-metal environment, you’ll know that large-scale testing comes with serious challenges. Most of us don’t have access to enough physical servers—or even virtual machines—to simulate the kinds of large-scale environments we need for stress testing, especially when deploying hundreds...

#### Ironic & Ironic Python Agent (IPA)

- Ironic is a bare-metal provisioning tool that handles tasks like booting
servers, deploying bootable media (e.g., operating systems) to disk, and
configuring hardware. Think of Ironic as the piece of software that manages
actual physical servers. In a Metal3 workflow, Ironic receives orders from BMO
and translates them into actionable steps. Ironic has multiple ways to interact
with the machines, and one of them is the so-called " agent-based direct deploy"
method, which is commonly used by BMO. The agent mentioned is called __Ironic
Python Agent__ (IPA), which is a piece of software that runs on each bare-metal
node and carries out Ironic's instructions. It interacts with the hardware
directly, like wiping disks, configuring networks, and handling boot processes.

In a typical Metal3 workflow, BMO reads the desired state of the node from the
BMH object, translates the Kubernetes reconciling logic to concrete actions, and
forwards them to Ironic, which, as part of the provisioning process, tells IPA
the exact steps it needs to perform to get the nodes to desired states. During
the first boot after node image installation,

> [cloud-init](https://cloud-init.io/)) will install Kubernetes components on the

Reviewer 2: ```suggestion
[cloud-init](https://cloud-init.io/) will install Kubernetes components on the
```
I would personally avoid putting so much details here because many of them may not always be correct. Cloud-init *can* be used, but so can ignition, cloud-init *may* install kubernetes components, or they may be already in the image, etc...

powerful, consider reducing the nodes you want to provision. Something like 100
nodes will require minimal resources and time while still being impressive.

### Step 2: Install BMO and Ironic

In Metal3's typical workflow, we usually rely on Kustomize to install Ironic and
BMO. Kustomize helps us define configurations for Kubernetes resources, making
it easier to customize and deploy services. However, our current Kustomize
overlay for Metal3 configures only a single Ironic instance. This setup works
well for smaller environments, but it becomes a bottleneck when scaling up and
handling thousands of nodes.

That's where Ironic's __special mode__ comes into play. Ironic has the ability
to run __multiple Ironic conductors__ while sharing the same database. The best
part? Workload balancing between conductors happens automatically, which means
that no matter which Ironic conductor receives a request, the load is evenly
distributed across all conductors, ensuring efficient provisioning. Achieving
this requires separating __ironic conductor__ from the database, which allows us
to scale up the conductor part. Each __conductor__ will have its own

> `PROVISIONING_IP,` hence the need to have a specialized `configMap.`

Reviewer 2: ```suggestion
`PROVISIONING_IP`, hence the need to have a specialized `configMap.`
```

We used [Helm](https://helm.sh/) for this purpose. In our Helm chart, the
__Ironic conductor__ container and __HTTP server (httpd)__ container are
separated into a new pod, and the rest of the ironic package (mostly

> `MariaDB`-ironic database) stays in another pod. A list of PROVISIONING_IPs is

Reviewer 2: This looks quite strange rendered when one part of the word is in ticks and the other outside. I would just skip the ticks. They aren't usually used for product names anyway.
```suggestion
MariaDB-ironic database) stays in another pod. A list of PROVISIONING_IPs is
```

provided by the chart's `values.yaml`, and for each IP, an  __ironic conductor__
pod is created, along with a config map whose values are rendered with the IP's
value. This way, we can dynamically scale up/down ironic (or, more specifically,
__ironic conductors__) by simply adding/removing ips.

Another piece of information that we need to keep in mind is the ipa-downloader
container. In our current metal3-dev-env, the IPA-downloader container runs as

> an init Container for ironic, and its job is to download the IPA image from its
> satisfactory to a Persistent Volume. This image contains the __Ironic Python

Reviewer 2: ```suggestion
an init Container for ironic, and its job is to download the IPA image
to a Persistent Volume. This image contains the __Ironic Python
```

Agent__, and it is assumed to exist by Ironic. For the multiple-conductor
scenario, running the same init-container for all the conductors, at the same
time, could be slow and/or fail due to network issue. To make it work, we made a
small "hack" in the chart: the ipa image will exist in a specific location
inside the minikube host, and all the conductor pods will mount to that same
location. In production, a more throughout solution might be to keep the
IPA-downloader as an init-container, but points the image to the local image
server, which we set up in the previous step.

BMO, on the other hand, still works well with kustomize, as we do not need to
scale it. As with typical metal3 workflow, BMO and Ironic must share some
authentication to work with TLS.

You can check out the full Ironic helm chart

> [here](https://github.com/Nordix/metal3-clusterapi-docs/tree/main/Support/Multitenancy/Scalability-with-Fake-Nodes/ironic)

Reviewer 2: ```suggestion
[here](https://github.com/Nordix/metal3-clusterapi-docs/tree/main/Support/Multitenancy/Scalability-with-Fake-Nodes/ironic).
```

out if you need an idea). Still, we must store this information to generate the
BMH objects that match those "nodes." The `ip` is, on the other hand, not
essential. It could be anything.

We must also start up the __sushy-tools__ container in this step. It is a tool
that simulates the [Baseboard Management
Controller](https://www.techtarget.com/searchnetworking/definition/baseboard-management-controller)
for non-bare-metal hardware, and we have been using it extensively inside Metal3
dev-env and CI to control and provision VMs as if they are bare-metal nodes. In
a bare-metal setup, Ironic will ask the BMC to install IPA on the node, and in
our setup, __sushy-tools__ will get the same request, but it will simply fake
the installation and, in the end, forward __Ironic__ traffic to the __Fake IPA__
container.

Another piece of information we will need is the cert that __Ironic__ will use
in its communication with __IPA__. IPA is supposed to get it from Ironic, but as
__Fake IPA__ cannot do that (at least not yet), we must get the cert and provide
it in __Fake IPA__ config.

```shell

> 
> mkdir cert
> kubectl get secret -n baremetal-operator-system ironic-cert -o json -o=jsonpath="{.data.ca\.crt}" | base64 -d >cert/ironic-ca.crt

Reviewer 2: Remove the blank line and consider breaking up the long line.
```suggestion
mkdir cert
kubectl get secret -n baremetal-operator-system \
  ironic-cert -o json -o=jsonpath="{.data.ca\.crt}" \
  | base64 -d >cert/ironic-ca.crt
```

```

Also note that one set of __sushy-tools__ and __Fake IPA__ containers won't be
enough to provision 1000 nodes. Just like __Ironic__, they need to be scaled up
extensively (about 20-30 pairs will be sufficient for 1000 nodes), but
fortunately, the scaling is straightforward: We just need to give them different
ports. Both of these components also require a Python-based config file. For
convenience, in this setup, we create a big file and provide it to both of them,
using the following shell script:

```shell

> 

Reviewer 2: ```suggestion
```

for i in $(seq 1 "$N_SUSHY"); do
  container_conf_dir="$SUSHY_CONF_DIR/sushy-$i"
  fake_ipa_port=$((9901 + (($i % ${N_FAKE_IPA:-1}))))
  port=$((8000 + i))
  ports+=(${port})
  ports+=(${fake_ipa_port})
  mkdir -p "${container_conf_dir}"
  cat <<'EOF' >"${container_conf_dir}"/htpasswd

> admin:$2b$12$/dVOBNatORwKpF.ss99KB.vESjfyONOxyH.UgRwNyZi1Xs/W2pGVS

Reviewer 2: A comment explaining this would be good

the new cluster's namespace and cluster name), and it will give us an IP and a
port, which we can plug into our cluster template and then run `clusterctl
generate cluster`.

Under the hood, __FKAS__ generates unique API servers for different clusters.
Each of the fake API servers does the following jobs:

- Mimicking API Calls: The Fake Kubernetes API server was set up to respond to
the essential Kubernetes API calls made during provisioning.
- Node Registration: When CAPM3 registered nodes, the Fake API server returned
success responses, making Metal3 believe the nodes had joined a real Kubernetes
cluster.
- Cluster Health and Status: The Fake API responded with "healthy" statuses,
allowing CAPI/CAPM3 to continue its workflow without interruption.
- Node Creation and Deletion: When CAPI queried for node status or attempted to
add/remove nodes, the Fake API server responded realistically, ensuring the
provisioning process continued smoothly.
- Pretending to Host Kubelet: The Fake API server also simulated kubelet
responses, which allowed CAPI/CAPM3 to interact with the fake clusters as though
they were managing actual nodes.

> Note that in this experiment, we provisioned every of the 1000 fake nodes to a

Reviewer 2: ```suggestion
Note that in this experiment, we provisioned every one of the 1000 fake nodes to a
```

have now will be a combination of a port with FKAS's IP address, which isn't
accessible from outside the cluster. What we should do now is:

- Edit the `kubeconfig-<cluster-name>.yaml` so that the endpoint is in the form
`localhost:<port>`.
- Port-forward the FKAS Pod to the same port the kubeconfig has shown.

And voila, now we can access the fake cluster with `kubectl --kubeconfig
kubeconfig-<cluster-name>.yaml`. You can inspect its state and check the
resources (nodes, pods, etc.), but we won't be able to run any workload on it as
it's fake.

## Results

In this post, we have demonstrated how it is possible to "generate"
bare-metal-based Kubernetes clusters from thin air (or rather, a bunch of nodes
that do not exist). Of course, these "clusters" are not very useful. Still,
successfully provisioning them without letting any of our main components
(__CAPI__, __CAPM3__, __BMO__, and __Ironic__) know they are working with fake
hardware proves that __Metal3__ is capable of handling a heavy workload and
provision multiple nodes/clusters.

If interested, you could also check (and try out) the experiment by yourself

> [here](https://github.com/Nordix/metal3-clusterapi-docs/blob/main/Support/Multitenancy/Scalability-with-Fake-Nodes/README.md)

Reviewer 2: ```suggestion
[here](https://github.com/Nordix/metal3-clusterapi-docs/blob/main/Support/Multitenancy/Scalability-with-Fake-Nodes/README.md).
```

- Ironic is a bare-metal provisioning tool that handles tasks like booting
servers, deploying bootable media (e.g., operating systems) to disk, and
configuring hardware. Think of Ironic as the piece of software that manages
actual physical servers. In a Metal3 workflow, Ironic receives orders from BMO
and translates them into actionable steps. Ironic has multiple ways to interact
with the machines, and one of them is the so-called " agent-based direct deploy"
method, which is commonly used by BMO. The agent mentioned is called __Ironic
Python Agent__ (IPA), which is a piece of software that runs on each bare-metal
node and carries out Ironic's instructions. It interacts with the hardware
directly, like wiping disks, configuring networks, and handling boot processes.

In a typical Metal3 workflow, BMO reads the desired state of the node from the
BMH object, translates the Kubernetes reconciling logic to concrete actions, and
forwards them to Ironic, which, as part of the provisioning process, tells IPA
the exact steps it needs to perform to get the nodes to desired states. During
the first boot after node image installation, Kubernetes components will be

> installed on the nodes, and once the process succeeds, Ironic passes control

Reviewer: ```suggestion
installed on the nodes by cloud-init, and once the process succeeds, Ironic passes control
```

> installed on the nodes, and once the process succeeds, Ironic passes control

Reviewer: Leave out this ironic passing control and stuff. 

Ironic doesn't do any sort of handover, it doesn't even know about K8s or cloud-init.  Instead mention that after reboot the IPA is no longer running and Ironic only have power control and BMC access nothing else.

it in __Fake IPA__ config.

```shell
mkdir cert
kubectl get secret -n baremetal-operator-system ironic-cert -o json \
  -o=jsonpath="{.data.ca\.crt}" | base64 -d >cert/ironic-ca.crt
```

Also note that one set of __sushy-tools__ and __Fake IPA__ containers won't be
enough to provision 1000 nodes. Just like __Ironic__, they need to be scaled up
extensively (about 20-30 pairs will be sufficient for 1000 nodes), but
fortunately, the scaling is straightforward: We just need to give them different
ports. Both of these components also require a Python-based config file. For
convenience, in this setup, we create a big file and provide it to both of them,
using the following shell script:

```shell
for i in $(seq 1 "$N_SUSHY"); do
  container_conf_dir="$SUSHY_CONF_DIR/sushy-$i"

  # Use round-robin to choose fake-ipa and sushy-tools containers for the node
  fake_ipa_port=$((9901 + (($i % ${N_FAKE_IPA:-1}))))
  sushy_tools_port=$((8000 + i))
  ports+=(${sushy_tools_port})

  # This is only so that we have the list of the needed ports for other
  # purposes, like configuring the firewalls.
  ports+=(${fake_ipa_port})

  mkdir -p "${container_conf_dir}"

>   # Use default username (`username`) and password (`password`) in base64 

Reviewer 2: Sorry but this is not correct. It is an htpasswd file and the username is `admin` (first part of the string). The rest is probably a bcrypt hash of the password, which I don't know what it is. Anyway you can probably just mention that this an htpasswd file used for securing the container(s)? I'm not sure if it is used for sushy-tools or fake IPA or both though.
